torch/csrc/jit/script/module.h

```diff
@@ -381,120 +393,291 @@ struct TORCH_API Module {
+// but skip modules, buffers, and other attributes.
+// See ModulePolicy for comments about Policy object's API.
+template <typename Policy>
 struct TORCH_API slot_iterator_impl {
-  slot_iterator_impl(Module module, c10::optional<EntityType> type, size_t i)
-      : module_(module), type_(type), i_(i) {
-    advance_to_valid();
-  }
-  T operator*() const;
-  T operator->() const {
+  using SlotCursor = detail::SlotCursor;
+  using value_type = typename Policy::value_type;
+  slot_iterator_impl(
+      Module root,
+      bool recurse, // if true, do a depth-first search, otherwise, just look at
+                    // slots of root
+      bool return_module) // if true include root itself as the first thing
+                          // visited (used in modules())
+      : cursors_({SlotCursor{root, return_module ? -1 : 0}}),
+        recurse_(recurse) {
+    // advance iterator to first valid element (or the end, if empty)
+    while_not_valid_next();
+  }
+  // empty frame_, represents end of iteration
+  slot_iterator_impl() : recurse_(false) {}
+  value_type operator*() const {
+    return Policy::create(cursors_, cur());
+  }
+  value_type operator->() const {
     return **this;
   }
   slot_iterator_impl& operator++() {
-    ++i_;
-    advance_to_valid();
+    next_valid();
     return *this;
   }
   slot_iterator_impl operator++(int) {
+    // this is really expensive, should we delete it so people don't use it
+    // instead of prefix?
     slot_iterator_impl old = *this;
     ++(*this);
     return old;
   }
 
  private:
-  void advance_to_valid() {
-    while (i_ < module_.num_slots() &&
-           (type_ && module_.entity_type(i_) != *type_)) {
-      ++i_;
+  // return_module() is a corner case where instead of returning a submodule
+  // of root, we are return root itself, because we are iterating modules(),
+  // which contains the root module itself.
+  // It is represented with a single Frame whose index is -1.
+  bool return_module() const {
+    return top().i_ == -1;
+  }
+  const SlotCursor& top() const {
+    return cursors_.back();
+  }
+  SlotCursor& top() {
+    return cursors_.back();
+  }
+  IValue cur() const {
+    return return_module() ? top().module_.module_object()
+                           : top().module_.module_object()->getSlot(top().i_);
+  }
+
+  // advance to the next slot in a depth first pre-order traversal of the
+  // modules slots. This function does not guarentee the next slot is a
+  // valid element of the iteration. That is done by valid().
+  // invariant: !cursors_.empty()
+  void next() {
+    // we just returned the module itself, advance i_ to 0 so we are now
+    // at the first slot of the module.
+    if (return_module()) {
+      ++top().i_;
+      return;
+    }
+    // the last traversal action advanced beyond the number of slots in the
+    // module so continue the iteration in the parent.
+    if (top().i_ >= int64_t(top().module_.num_slots())) {
```

Comment:
Should we do it *while* we're beyond the number of slots instead of *if*? I.e. shall we pop back until we reach a valid position or do we intentionally want to pop back only once?

Reply:
That would be correct too, but I went with this way to make the components easier to understand: next() does 1 step, while_not_valid is responsible for repeating.